lib/app/endpoint/oauth2_endpoints.rb

```diff
@@ -5,91 +5,74 @@ class App
+
+              halt 500, message
+            elsif @instance&.waiting_on_sequence&.wait?
+              @error_message = "State provided in redirect (#{params[:state]}) does not match expected state (#{@instance.state})."
+            else
+              redirect "#{base_path}/#{cookies[:instance_id_test_set]}/?error=no_state&state=#{params[:state]}"
+            end
+          end
+
+          get '/oauth2/:key/launch/?' do
+            @instance = Inferno::Models::SequenceResult.recent_results_for_iss(params[:iss]).testing_instance
+            pass if @instance.present?
+
+            @instance = Inferno::Models::TestingInstance.get(instance_id_from_cookie)
+            if @instance.nil?
+              message = "Error: No actively running launch sequences found for iss #{params[:iss]}. " \
+                        'Please ensure that the EHR launch test is actively running before attempting to launch Inferno from the EHR.'
+              halt 500, message
+            elsif @instance.waiting_on_sequence&.wait?
+              @error_message = 'No iss for redirect'
+            else
+              redirect "#{base_path}/#{cookies[:instance_id_test_set]}/?error=no_ehr_launch&iss=#{params[:iss]}"
             end
-            halt 500, 'Error: Could not find a running test that match this set of critera' unless !instance.nil? &&
-                                                                                                   instance.client_endpoint_key == params[:key] &&
-                                                                                                   %w[launch redirect].include?(params[:endpoint])
+          end
 
-            sequence_result = instance.waiting_on_sequence
+          get '/oauth2/:key/:endpoint/?' do
+            halt 500, 'Error: Could not find a running test that match this set of criteria' unless running_test_found?
+
+            sequence_result = @instance.waiting_on_sequence
             if sequence_result&.wait?
-              test_set = instance.module.test_sets[sequence_result.test_set_id.to_sym]
+              test_set = @instance.module.test_sets[sequence_result.test_set_id.to_sym]
               failed_test_cases = []
               all_test_cases = []
               test_case = test_set.test_case_by_id(sequence_result.test_case_id)
               test_group = test_case.test_group
 
-              client = FHIR::Client.new(instance.url)
-              case instance.fhir_version
-              when 'stu3'
-                client.use_stu3
-              when 'dstu2'
-                client.use_dstu2
-              else
-                client.use_r4
-              end
-              client.default_json
-              sequence = test_case.sequence.new(instance, client, settings.disable_tls_tests, sequence_result)
+              client = FHIR::Client.for_testing_instance(@instance)
```

Comment:
Would it be better to have something like a `get_client` method on SequenceBase  or TestingInstance instead of patching FHIR::Client?